mypy/checker.py

```diff
@@ -1062,6 +1065,15 @@ def check_func_def(self, defn: FuncItem, typ: CallableType, name: Optional[str])
 
             self.binder = old_binder
 
+    def check_unbound_return_typevar(self, typ: CallableType) -> None:
+        """Fails when the return typevar is not defined in arguments."""
+        arg_type_visitor = CollectArgTypes()
+        for argtype in typ.arg_types:
+            argtype.accept(arg_type_visitor)
+
+        if typ.ret_type not in arg_type_visitor.arg_types and typ.ret_type in typ.variables:
```

Comment:
Could we micro optimise this by first checking `if typ.ret_type in typ.variables` and only then doing the collection of arg types?

Reply:
Good idea!